Enable screenshots of animated components

allenchen1154 opened the pull request:
This adds support for capturing animated PNGs of components annotated with `@ShowkaseComposable`, by specifying the new `screenshotCaptureType` parameter. For example, to capture an animation for 2 seconds with a framerate of 60fps:
```kotlin
@ShowkaseComposable(
    name = "MyAnimatedComponent",
    group = "AnimatedGroup",
    defaultStyle = true,
    screenshotCaptureType = ScreenshotCaptureType.SingleAnimatedImage,
    captureDurationMillis = 2000,
    captureFramerate = 60,
)
@Composable
fun MyAnimatedComponentPreview_Default() {
    var animatedComponentState by remember { mutableStateOf(...) }

    LaunchedEffect(Unit) {
        animatedComponentState = ...
    }

    MyAnimatedComponent(animatedComponentState)
}

```

Updates Paparazzi to `2.0.0-alpha02` to fix [this issue](https://github.com/cashapp/paparazzi/issues/1877) with compileSdk 36.
 

https://github.com/user-attachments/assets/d05f8cb7-6876-43bd-aa03-0b530ee89cee

allenchen1154 committed:
Enable screenshots of animated components

This adds support for capturing animated PNGs of components annotated with `@ShowkaseComposable`, by specifying the new `screenshotConfig` parameter. For example, to capture an animation for 2 seconds with a framerate of 60fps:
```kotlin
@ShowkaseComposable(
    name = "MyAnimatedComponent",
    group = "AnimatedGroup",
    defaultStyle = true,
    screenshotCaptureType = ScreenshotCaptureType.SingleAnimatedImage,
    captureDurationMillis = 2000,
    captureFramerate = 60,
)
@Composable
fun MyAnimatedComponentPreview_Default() {
    var animatedComponentState by remember { mutableStateOf(...) }

    LaunchedEffect(Unit) {
        animatedComponentState = ...
    }

    MyAnimatedComponent(animatedComponentState)
}

```

Updates Paparazzi to `2.0.0-alpha02` to fix [this issue](https://github.com/cashapp/paparazzi/issues/1877) with compileSdk 36.

allenchen1154 commented on showkase-processor/src/main/java/com/airbnb/android/showkase/processor/writer/PaparazziShowkaseScreenshotTestWriter.kt:
```diff
@@ -22,7 +22,7 @@ class PaparazziShowkaseScreenshotTestWriter(private val environment: XProcessing
         testClassName: String
     ) {
-        val showkaseScreenshotTestClassName = "${testClassName}_PaparazziShowkaseTest"
+        val showkaseScreenshotTestClassName = "${testClassName}Impl"
```
Note: This contains the fixes in https://github.com/airbnb/Showkase/pull/419 to shorten the generated Paparazzi image filenames - due to changes in Paparazzi, we can now encounter `java.io.FileNotFoundException: File name too long` on some systems.

allenchen1154 committed:
Detekt formatting

allenchen1154 committed:
Regenerate golden screenshots for name changes

vinaygaba commented on showkase-annotation/src/main/java/com/airbnb/android/showkase/annotation/ScreenshotConfig.kt:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Configuration for how screenshots of the annotated Composable should be captured.
+ */
+sealed interface ScreenshotConfig {
```
Nice!

vinaygaba commented on showkase-annotation/src/main/java/com/airbnb/android/showkase/annotation/ShowkaseComposable.kt:
```diff
@@ -74,4 +74,52 @@ annotation class ShowkaseComposable(
+     * Used by Paparazzi snapshot testing to determine if the component has any animation, and how to capture
+     * the screenshot.
+     */
+    val screenshotCaptureType: ScreenshotCaptureType = ScreenshotCaptureType.SingleStaticImage,
```
Wish there was a cleaner way (with a single annotation param) to do this so that it was more encapsulated. I think Arrays are the only choice but that's ugly.

allenchen1154 replied:
Yeah unfortunately we're limited by what annotation classes support, hence the need to convert it to `ScreenshotConfig` later down the line.

elihart replied:
@allenchen1154 can you use nested annotations to encapsulate all animation related properties?

eg
```
annotation class Foo(val bar: Bar)

annotation class Bar(val x: Int)

@Foo(bar = Bar(x = 1))
class Baz
```

vinaygaba replied:
What @elihart is suggesting would be really neat!

allenchen1154 replied:
@elihart thanks for the suggestion, moved these params into a wrapper annotation

vinaygaba commented:
What's the current behavior when it comes to the Showkase browser itself? How are those preview rendered in the browser? Additionally, if you are using offsets, should they show up as separate previews? I think that would be the ideal behavior but since you are using Paparazzi for the screenshots, I suspect that might not be possible without some additional work as the browser shows live composable instead.

vinaygaba commented:
In an ideal case, we maintain parity between what you see in the browser and what gets screenshot tested.

allenchen1154 commented:
@vinaygaba Currently the previews that show in the Showkase browser would just play the animation as it's specified by the preview Composable. So in the example I added, it would simply animate the red square to the right one time. You're right that generating separate browser components for the offsets would require more work/codegen. I think having this discrepancy is OK for now, as I'm hoping the APNG approach will be preferred.

allenchen1154 commented:
@vinaygaba (The offset approach doesn't actually work right now due to a bug in Paparazzi: https://github.com/cashapp/paparazzi/pull/1645)

allenchen1154 committed:
Fix more detekt formatting errors

allenchen1154 committed:
Use 10fps instead of 30fps as default

allenchen1154 committed:
Update processor testing output files

allenchen1154 committed:
Update `dev.zacsweers.kctfork` library

This was causing an error in the compilation tests when running in KAPT mode.

allenchen1154 committed:
Fix detekt line length

allenchen1154 committed:
Fix detekt formatting

vinaygaba reviewed (approved):
Changes LGTM. Would be nice to add test coverage - processor tests especially

allenchen1154 committed:
Add processor test

This test checks that when `screenshotCaptureType` is provided, the generated`ShowkaseBrowserComponent` instances have the correct `ScreenshotConfig` values.

allenchen1154 commented:
@vinaygaba I added a processor test, will merge and prepare a 1.1.0 release.

allenchen1154 committed:
Move screenshot configuration values into a wrapper annotation class

allenchen1154 committed:
Update naming

allenchen1154 committed:
Fix long function detekt error

allenchen1154 committed:
Remove incorrect framerate on sample

allenchen1154 committed:
Add default value for nested annotation
